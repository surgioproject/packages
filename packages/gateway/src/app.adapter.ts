import { ExpressAdapter } from '@nestjs/platform-express'
import express from 'express'

export function createAdapter(): ExpressAdapter {
  const instance = express()
  return new ExpressAdapter(instance)
}
