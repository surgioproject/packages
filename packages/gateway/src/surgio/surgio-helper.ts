import os from 'os'
import { basename, join } from 'path'
import fs from 'fs-extra'
import { Environment } from 'nunjucks'
import semver from 'semver'
import { Logger } from '@nestjs/common'
import { getEngine } from 'surgio/generator'
import { getProvider, PossibleProviderType } from 'surgio/provider'
import type {
  ArtifactConfig,
  CommandConfig,
  RemoteSnippet,
} from 'surgio/internal'
import { packageJson as corePackageJson, cleanCaches } from 'surgio/internal'
import { TMP_FOLDER_NAME } from 'surgio/constant'
import { createHash } from 'crypto'

export const KEY = 'SURGIO_HELPER'

export class SurgioHelper {
  public remoteSnippetList?: ReadonlyArray<RemoteSnippet>
  public artifactList: ReadonlyArray<ArtifactConfig>
  public providerMap: Map<string, PossibleProviderType> = new Map()
  public readonly templateEngine: Environment
  public readonly configHash: string

  constructor(public cwd: string, public readonly config: CommandConfig) {
    this.artifactList = config.artifacts
    this.templateEngine = getEngine(config.templateDir, {
      clashCore: config.clashConfig?.clashCore,
    })
    this.configHash = this.getConfigSHA256Hash()
  }

  public async init(): Promise<this> {
    await this.checkCoreVersion()
    await this.readProviders()

    return this
  }

  private async readProviders(): Promise<void> {
    const files = await fs.readdir(this.config.providerDir, {
      encoding: 'utf8',
    })

    async function readProvider(
      path: string
    ): Promise<PossibleProviderType | undefined> {
      const provider = await (() => {
        try {
          if (!path.endsWith('.js')) {
            return undefined
          }

          const providerName = basename(path, '.js')

          return getProvider(providerName, require(path))
        } catch (err) {
          Logger.error(`读取 Provider (${path}) 失败: ` + err.message)
          return undefined
        }
      })()

      if (!provider?.type) {
        return undefined
      }

      return provider
    }

    for (const file of files) {
      const result = await readProvider(join(this.config.providerDir, file))
      if (result) {
        this.providerMap.set(result.name, result)
      }
    }
  }

  private async checkCoreVersion(): Promise<void> {
    const gatewayPkgFile = require('../../package.json')
    const peerVersion = gatewayPkgFile.peerDependencies.surgio
    const corePkgVersion = corePackageJson.version as string

    // Pre-release doesn't need to check
    if (corePkgVersion.includes('-')) {
      return
    }

    if (!semver.satisfies(corePkgVersion, peerVersion)) {
      Logger.warn('', undefined, false)
      Logger.warn(
        'Surgio 版本过低，请运行下面命令升级后重新运行！',
        undefined,
        false
      )
      Logger.warn(
        `要求版本 ${peerVersion}，当前版本 ${corePackageJson.version}`,
        undefined,
        false
      )
      Logger.warn('', undefined, false)
      Logger.warn('  命令：', undefined, false)
      Logger.warn('  npm install surgio@latest', undefined, false)
      Logger.warn('', undefined, false)
      throw new Error('Surgio 版本过低')
    }
  }

  private getConfigSHA256Hash(): string {
    return createHash('sha256')
      .update(JSON.stringify(this.config))
      .digest('hex')
  }

  public async cleanCache(): Promise<void> {
    const tmpDir = join(os.tmpdir(), TMP_FOLDER_NAME)

    if (this.remoteSnippetList) {
      Logger.log('已清除远程片段')
      this.remoteSnippetList = undefined
    }

    if (fs.existsSync(tmpDir)) {
      Logger.log('已清除文件缓存')
      await fs.remove(tmpDir)
    }

    await cleanCaches()
    Logger.log('已清除 内存/Redis 缓存')
  }
}
