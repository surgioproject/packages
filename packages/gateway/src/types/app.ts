import { Request } from 'express'

import { Role } from '../constants/role'

export interface UserContext {
  readonly accessToken: string
  readonly roles: Array<Role>
}

export interface EnrichedRequest extends Request {
  user?: UserContext
}
