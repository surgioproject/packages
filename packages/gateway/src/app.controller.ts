import {
  Controller,
  Get,
  Res,
  Param,
  Query,
  HttpException,
  HttpStatus,
  UseGuards,
  Req,
  Logger,
} from '@nestjs/common'
import { Request, Response } from 'express'
import { Artifact } from 'surgio/generator'
import _ from 'lodash'
import { getUrl } from 'surgio/utils'
import { URL } from 'url'
import dayjs from 'dayjs'
import duration from 'dayjs/plugin/duration'
import NodeCache from 'node-cache'

import { APIAuthGuard } from './auth/api-auth.guard'
import { Roles } from './auth/roles.decorator'
import { Role } from './constants/role'
import { SurgioService } from './surgio/surgio.service'

dayjs.extend(duration)

const resCache = new NodeCache({
  maxKeys: 100,
  stdTTL: dayjs.duration({ days: 7 }).asSeconds(),
})

@Controller()
@UseGuards(APIAuthGuard)
export class AppController {
  private readonly logger = new Logger(AppController.name)

  constructor(private readonly surgioService: SurgioService) {}

  @Get('/get-artifact/:name')
  @Roles(Role.VIEWER)
  public async getArtifact(
    @Res() res: Response,
    @Param() params: { readonly name: string },
    @Query() query: GetArtifactQuery,
    @Req() req: Request
  ): Promise<void> {
    const dl = query.dl
    const format = query.format
    const filter = query.filter
    const urlParams = _.omit(query, ['dl', 'format', 'filter', 'access_token'])
    const artifactName: string = params.name
    const userAgent = req.headers['user-agent']
    let artifact: string | undefined | Artifact
    let isCache = false

    try {
      artifact =
        format !== void 0
          ? await this.surgioService.transformArtifact(
              artifactName,
              format,
              filter
            )
          : await this.surgioService.getArtifact(artifactName, {
              downloadUrl: new URL(
                req.url,
                this.surgioService.config.publicUrl
              ).toString(),
            })
    } catch (err) {
      if (resCache.has(req.url)) {
        isCache = true
        artifact = resCache.get(req.url) as string

        this.logger.warn('Artifact 生成错误，使用缓存')
        this.logger.warn(err.stack || err)
      } else {
        throw err
      }
    }

    if (typeof artifact !== 'undefined') {
      res.header('content-type', 'text/plain; charset=utf-8')
      res.header('cache-control', 'private, no-cache, no-stores')

      if (dl === '1') {
        res.header(
          'content-disposition',
          `attachment; filename="${artifactName}"`
        )
      }

      this.logger.warn(
        `[download-artifact] ${artifactName} "${userAgent || '-'}"`
      )

      await this.sendPayload(
        req,
        res,
        artifact,
        {
          ...urlParams,
          userAgent: userAgent || '',
        },
        isCache
      )
    } else {
      throw new HttpException('NOT FOUND', HttpStatus.NOT_FOUND)
    }
  }

  @Get('/export-providers')
  @Roles(Role.VIEWER)
  public async exportProvider(
    @Req() req: Request,
    @Res() res: Response,
    @Query() query: ExportProviderQuery
  ): Promise<void> {
    const providers: string[] = query.providers
      ? query.providers.split(',').map((item) => item.trim())
      : []
    const userAgent = req.headers['user-agent']

    if (!providers.length) {
      throw new HttpException(
        '参数 provider 必须指定至少一个值',
        HttpStatus.BAD_REQUEST
      )
    }

    providers.forEach((provider) => {
      if (!this.surgioService.surgioHelper.providerMap.has(provider)) {
        throw new HttpException(
          `provider ${provider} 不存在`,
          HttpStatus.NOT_FOUND
        )
      }
    })

    const format = query.format
    const template = query.template

    if (!format && !template) {
      throw new HttpException(
        '参数 format 和 template 必须指定至少一个值',
        HttpStatus.BAD_REQUEST
      )
    }

    const dl = query.dl
    const filter = query.filter
    const urlParams = _.omit(query, [
      'dl',
      'format',
      'template',
      'filter',
      'access_token',
      'providers',
    ])
    let artifact: Artifact | string
    let isCache = false

    try {
      if (format) {
        artifact = await this.surgioService.exportProvider(
          providers[0],
          format,
          undefined,
          {
            filter,
            ...(providers.length > 1
              ? {
                  combineProviders: providers.splice(1),
                }
              : null),
          }
        )
      } else {
        artifact = await this.surgioService.exportProvider(
          providers[0],
          undefined,
          template,
          {
            filter,
            downloadUrl: new URL(
              req.url as string,
              this.surgioService.config.publicUrl
            ).toString(),
            ...(providers.length > 1
              ? {
                  combineProviders: providers.splice(1),
                }
              : null),
          }
        )
      }
    } catch (err) {
      if (resCache.has(req.url)) {
        isCache = true
        artifact = resCache.get(req.url) as string

        this.logger.warn('Provider 导出错误，使用缓存')
        this.logger.warn(err.stack || err)
      } else {
        throw err
      }
    }

    if (artifact instanceof Artifact) {
      res.header('content-type', 'text/plain; charset=utf-8')
      res.header('cache-control', 'private, no-cache, no-stores')

      if (dl === '1') {
        res.header(
          'content-disposition',
          `attachment; filename="${artifact.artifact.name}"`
        )
      }

      this.logger.warn(
        `[download-artifact] ${artifact.artifact.name} "${userAgent || '-'}"`
      )

      await this.sendPayload(
        req,
        res,
        artifact,
        {
          ...urlParams,
          userAgent: userAgent || '',
        },
        isCache
      )
    } else {
      await this.sendPayload(req, res, artifact, undefined, isCache)
    }
  }

  @Get('/render')
  @Roles(Role.VIEWER)
  public async renderTemplate(
    @Req() req: Request,
    @Res() res: Response,
    @Query() query: RenderTemplateQuery
  ): Promise<void> {
    const { template } = query
    const config = this.surgioService.config
    const gatewayConfig = config?.gateway
    const gatewayHasToken = !!gatewayConfig?.accessToken

    if (!template) {
      throw new HttpException(
        '参数 template 必须指定一个值',
        HttpStatus.BAD_REQUEST
      )
    }

    try {
      res.header('content-type', 'text/plain; charset=utf-8')
      res.header(
        'cache-control',
        `s-maxage=${60 * 60 * 24}, stale-while-revalidate`
      )

      const html = this.surgioService.surgioHelper.templateEngine.render(
        `${template}.tpl`,
        {
          downloadUrl: new URL(
            req.url as string,
            this.surgioService.config.publicUrl
          ).toString(),
          getUrl: (p: string) =>
            getUrl(
              config.publicUrl,
              p,
              gatewayHasToken ? gatewayConfig?.accessToken : undefined
            ),
        }
      )

      res.send(html)
    } catch (err) {
      if (err.message.includes('template not found')) {
        throw new HttpException('NOT FOUND', HttpStatus.NOT_FOUND)
      } else {
        throw err
      }
    }
  }

  private async sendPayload(
    req: Request,
    res: Response,
    artifact: string | Artifact,
    urlParams?: Record<string, string>,
    isCachedPayload?: boolean
  ): Promise<void> {
    const config = this.surgioService.config
    const gatewayConfig = config?.gateway

    if (typeof artifact === 'string') {
      if (gatewayConfig?.useCacheOnError && !isCachedPayload) {
        resCache.set(req.url, artifact)
      }

      if (isCachedPayload) {
        res.header('x-use-cache', 'true')
      }

      res.send(artifact)
    } else {
      // 只支持输出单个 Provider 的流量信息
      if (!isCachedPayload && artifact.providerMap.size === 1) {
        const providers = artifact.providerMap.values()
        const provider = providers.next().value

        if (provider.supportGetSubscriptionUserInfo) {
          try {
            const subscriptionUserInfo =
              await provider.getSubscriptionUserInfo()

            if (subscriptionUserInfo) {
              const values = ['upload', 'download', 'total', 'expire'].map(
                (key) => `${key}=${subscriptionUserInfo[key] || 0}`
              )

              res.header('subscription-userinfo', values.join('; '))
            }
          } catch (err) {
            this.logger.error('处理订阅信息失败')
            this.logger.error(err.stack || err)
          }
        }
      }

      const body = artifact.render(undefined, {
        ...(urlParams ? this.processUrlParams(urlParams) : undefined),
      })

      if (gatewayConfig?.useCacheOnError && !isCachedPayload) {
        resCache.set(req.url, body)
      }

      if (isCachedPayload) {
        res.header('x-use-cache', 'true')
      }

      res.send(body)
    }
  }

  private processUrlParams(
    urlParams: Record<string, string>
  ): Record<string, string> {
    const result: NonNullable<any> = Object.create(null)

    Object.keys(urlParams).forEach((key) => {
      _.set(result, key, urlParams[key])
    })

    return result
  }
}

interface GetArtifactQuery {
  readonly format: string
  readonly filter?: string
  readonly dl?: string
  readonly access_token?: string
  readonly [key: string]: string | undefined
}

interface RenderTemplateQuery {
  readonly template: string
}

interface ExportProviderQuery {
  readonly providers: string
  readonly format?: string
  readonly template?: string
  readonly filter?: string
  readonly dl?: string
  readonly access_token?: string
  readonly [key: string]: string | undefined
}
