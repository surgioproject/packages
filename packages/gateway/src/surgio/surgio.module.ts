import { Module, DynamicModule, Global } from '@nestjs/common'
import path from 'path'

import { loadConfig } from 'surgio/build/utils/config'
import { KEY, SurgioHelper } from './surgio-helper'
import { SurgioService } from './surgio.service'

@Global()
@Module({})
export class SurgioModule {
  public static register(options: { readonly cwd: string }): DynamicModule {
    return {
      module: SurgioModule,
      providers: [
        SurgioService,
        {
          provide: KEY,
          useFactory: (): Promise<SurgioHelper> => {
            const configFile = path.join(options.cwd, 'surgio.conf.js')
            const config = loadConfig(options.cwd, configFile)
            const helper = new SurgioHelper(options.cwd, config)

            return helper.init()
          },
        },
      ],
      exports: [SurgioService],
    }
  }
}
