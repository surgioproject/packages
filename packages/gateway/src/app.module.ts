import { MiddlewareConsumer, Module, RequestMethod } from '@nestjs/common'
import { ConfigModule, ConfigService } from '@nestjs/config'
import { join } from 'path'
import { ServeStaticModule } from '@nestjs/serve-static'
import getPackage from '@surgio/gateway-frontend'

import { AppController } from './app.controller'
import { ApiModule } from './api/api.module'
import { CookieParserMiddleware } from './middleware/cookie-parser.middleware'
import { PrepareMiddleware } from './middleware/prepare.middleware'
import { SurgioModule } from './surgio/surgio.module'
import { SurgioService } from './surgio/surgio.service'
import { AuthModule } from './auth/auth.module'
import configuration from './config/configuration'

const FE_MODULE = require.resolve('@surgio/gateway-frontend')
const frontendPackage = getPackage()

@Module({
  imports: [
    ServeStaticModule.forRoot({
      rootPath: join(FE_MODULE, '../build'),
      serveStaticOptions: {
        cacheControl: true,
        etag: true,
        maxAge: '31d',
        setHeaders: (res, path) => {
          if (
            path.endsWith('.js') ||
            path.endsWith('.css') ||
            path.endsWith('.json')
          ) {
            res.setHeader('x-frontend-version', frontendPackage.version)
          }
        },
      },
    }),
    ConfigModule.forRoot({
      ignoreEnvFile: true,
      isGlobal: true,
      load: [configuration],
    }),
    SurgioModule.register({
      cwd: process.env.SURGIO_PROJECT_DIR || process.cwd(),
    }),
    ApiModule,
    AuthModule,
  ],
  controllers: [AppController],
})
export class AppModule {
  constructor(private surgioService: SurgioService) {}

  configure(consumer: MiddlewareConsumer): void {
    const secret = this.surgioService.surgioHelper.configHash

    CookieParserMiddleware.configure(secret)
    consumer.apply(CookieParserMiddleware).forRoutes('*')

    consumer
      .apply(PrepareMiddleware)
      .exclude({ path: 'render', method: RequestMethod.ALL })
      .forRoutes(AppController)
  }
}
