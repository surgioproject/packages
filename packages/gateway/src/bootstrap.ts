import { NestFactory } from '@nestjs/core'
import { NestExpressApplication } from '@nestjs/platform-express'
import express from 'express'

import { AppModule } from './app.module'
import { AppExceptionsFilter } from './filter/app-exception.filter'
import { createAdapter } from './app.adapter'

export async function bootstrap(): Promise<NestExpressApplication> {
  const app = await NestFactory.create<NestExpressApplication>(
    AppModule,
    createAdapter(),
    {
      logger:
        process.env.NODE_ENV === 'production' ? ['error', 'warn'] : ['verbose'],
    }
  )

  applyMiddlwares(app)

  return app
}

export function applyMiddlwares(
  app: NestExpressApplication
): NestExpressApplication {
  app.useGlobalFilters(new AppExceptionsFilter())
  app.use(express.json())

  return app
}
