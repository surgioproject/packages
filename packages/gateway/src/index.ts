import 'source-map-support/register'
import { ConfigService } from '@nestjs/config'
import { createServer, IncomingMessage, OutgoingMessage, Server } from 'http'
import serverless, { Handler } from 'serverless-http'

import { bootstrap } from './bootstrap'

export const createHttpServer = (): Server => {
  const prepare = bootstrap().then((nestApp) => {
    return nestApp.init()
  })

  return createServer((req: IncomingMessage, res: OutgoingMessage) => {
    prepare.then((nestApp) => {
      nestApp.getHttpServer().emit('request', req, res)
    })
  })
}

export const startServer = (): Promise<Server> => {
  let port: number

  return bootstrap()
    .then((app) => {
      const configService = app.get<ConfigService>(ConfigService)
      port = configService.get('port') as number

      return app.listen(port)
    })
    .then((httpServer: Server) => {
      console.log('> Your app is ready at http://127.0.0.1:' + port)
      return httpServer
    })
    .catch((err) => {
      console.error('⚠️ 服务启动失败')
      console.error(err)
      process.exit(1)
    })
}

export const createLambdaHandler = () => {
  let handler: Handler

  return async (event, context) => {
    if (!handler) {
      const nestApp = await bootstrap()
      await nestApp.init()
      const adapter = nestApp.getHttpAdapter().getInstance()
      handler = serverless(adapter)
    }

    return await handler(event, context)
  }
}

export const bootstrapServer = bootstrap
