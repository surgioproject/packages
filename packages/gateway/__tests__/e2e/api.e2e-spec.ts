import { NestFactory } from '@nestjs/core'
import supertest from 'supertest'
import { NestExpressApplication } from '@nestjs/platform-express'

import { AppModule } from '../../src/app.module'
import { applyMiddlwares } from '../../src/bootstrap'
import { SurgioService } from '../../src/surgio/surgio.service'
import { extractCookies } from '../helper'

describe('ApiController (e2e)', () => {
  let app: NestExpressApplication
  let token
  let viewerToken
  let tokenCookie
  let surgioService: SurgioService

  beforeAll(async () => {
    app = await NestFactory.create(AppModule, { logger: false })
    applyMiddlwares(app)

    surgioService = app.get<SurgioService>(SurgioService)
    token = surgioService.config.gateway?.accessToken
    viewerToken = surgioService.config.gateway?.viewerToken

    await app.init()

    const auth = await supertest(app.getHttpServer()).post('/api/auth').send({
      accessToken: token,
    })
    const cookies = extractCookies(auth.header)
    tokenCookie = cookies._t.value
  })

  afterAll(async () => {
    await app.close()
  })

  test('/api/auth (POST)', async () => {
    const auth = await supertest(app.getHttpServer()).post('/api/auth').send({
      accessToken: token,
    })
    const cookies = extractCookies(auth.header)
    expect(cookies._t.value).toBeDefined()
    expect(Number(cookies._t.flags['Max-Age'])).toBe(60 * 60 * 24 * 31)
    expect(cookies._t.flags.HttpOnly).toBe(true)
  })

  test('/api/auth/validate-cookie (GET)', async () => {
    await supertest(app.getHttpServer())
      .post('/api/auth/validate-cookie')
      .set('Cookie', `_t=${tokenCookie}`)
      .expect(200)

    await supertest(app.getHttpServer())
      .post('/api/auth/validate-cookie')
      .set('Cookie', `_t=wrong`)
      .expect(401)
  })

  test('/api/auth/validate-token (GET)', async () => {
    const res = await supertest(app.getHttpServer())
      .post('/api/auth/validate-token')
      .set('Authorization', `Bearer ${token}`)
      .expect(200)

    expect(res.body.data).toEqual({
      roles: ['admin', 'viewer'],
      viewerToken: 'efgh',
    })

    const viewerRes = await supertest(app.getHttpServer())
      .post('/api/auth/validate-token')
      .set('Authorization', `Bearer ${viewerToken}`)
      .expect(200)

    expect(viewerRes.body.data).toEqual({
      roles: ['viewer'],
      viewerToken: 'efgh',
    })

    await supertest(app.getHttpServer())
      .post('/api/auth/validate-token')
      .set('Authorization', `Bearer wrong`)
      .expect(401)
  })

  test('/api/config (GET)', async () => {
    const res = await supertest(app.getHttpServer())
      .get('/api/config')
      .expect(200)
    expect(res.body.data.needAuth).toBeDefined()
    expect(res.body.data.backendVersion).toBeDefined()
    expect(res.body.data.coreVersion).toBeDefined()
  })

  test('/api/artifacts (GET)', async () => {
    const res = await supertest(app.getHttpServer())
      .get('/api/artifacts')
      .set('Authorization', `Bearer ${token}`)
      .expect(200)

    expect(res.text).toMatchSnapshot()

    await supertest(app.getHttpServer())
      .get('/api/artifacts')
      .set('Authorization', `Bearer ${viewerToken}`)
      .expect(403)
  })

  test('/api/providers (GET)', async () => {
    const res = await supertest(app.getHttpServer())
      .get('/api/providers')
      .set('Authorization', `Bearer ${token}`)
      .expect(200)

    expect(res.text).toMatchSnapshot()
  })
})
