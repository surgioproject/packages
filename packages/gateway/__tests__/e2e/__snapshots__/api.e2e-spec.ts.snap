// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiController (e2e) /api/artifacts (GET) 1`] = `"{"status":"ok","data":[{"name":"test.conf","template":"test","provider":"ss","combineProviders":["custom","clash"]},{"name":"test2.conf","template":"test","provider":"rename"},{"name":"test3.conf","template":"test","provider":"subscription-userinfo"},{"name":"custom-params.conf","template":"custom-params","provider":"ss","combineProviders":["custom","clash"],"customParams":{"foo":"original","child":{"bar":"original"}}},{"name":"clash-error-after-first.conf","template":"test","provider":"clash-error-after-first"},{"name":"hooks.conf","template":"test","provider":"hooks"}]}"`;

exports[`ApiController (e2e) /api/providers (GET) 1`] = `"{"status":"ok","data":[{"name":"clash-error-after-first","type":"clash","url":"http://example.com/clash-sample-error-after-first.yaml","supportGetSubscriptionUserInfo":true},{"name":"clash","type":"clash","url":"http://example.com/clash-sample.yaml","supportGetSubscriptionUserInfo":true},{"name":"custom","type":"custom","supportGetSubscriptionUserInfo":false},{"name":"hooks","type":"custom","supportGetSubscriptionUserInfo":false},{"name":"rename","type":"shadowsocks_subscribe","url":"http://example.com/test-ss-sub.txt","supportGetSubscriptionUserInfo":true},{"name":"ss","type":"shadowsocks_subscribe","url":"http://example.com/test-ss-sub.txt","supportGetSubscriptionUserInfo":true},{"name":"subscription-userinfo","type":"shadowsocks_subscribe","url":"http://example.com/test-ss-sub-user-info.txt","supportGetSubscriptionUserInfo":true}]}"`;
